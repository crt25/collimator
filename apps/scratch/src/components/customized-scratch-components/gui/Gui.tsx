import classNames from "classnames";
import omit from "lodash.omit";
import {
  defineMessages,
  FormattedMessage,
  InjectedIntl,
  injectIntl,
} from "react-intl";
import { connect } from "react-redux";
import MediaQuery from "react-responsive";
import { Tab, Tabs, TabList, TabPanel } from "react-tabs";
import tabStyles from "react-tabs/style/react-tabs.css";
import Renderer from "scratch-render";
import VM from "scratch-vm";

import CostumeTab from "@scratch-submodule/scratch-gui/src/containers/costume-tab.jsx";
import SoundTab from "@scratch-submodule/scratch-gui/src/containers/sound-tab.jsx";
import Loader from "@scratch-submodule/scratch-gui/src/components/loader/loader.jsx";
import Box from "@scratch-submodule/scratch-gui/src/components/box/box.jsx";
import CostumeLibrary from "@scratch-submodule/scratch-gui/src/containers/costume-library.jsx";
import BackdropLibrary from "@scratch-submodule/scratch-gui/src/containers/backdrop-library.jsx";
import Watermark from "@scratch-submodule/scratch-gui/src/containers/watermark.jsx";

import Backpack from "@scratch-submodule/scratch-gui/src/containers/backpack.jsx";
import WebGlModal from "@scratch-submodule/scratch-gui/src/containers/webgl-modal.jsx";
import TipsLibrary from "@scratch-submodule/scratch-gui/src/containers/tips-library.jsx";
import Cards from "@scratch-submodule/scratch-gui/src/containers/cards.jsx";
import Alerts from "@scratch-submodule/scratch-gui/src/containers/alerts.jsx";
import DragLayer from "@scratch-submodule/scratch-gui/src/containers/drag-layer.jsx";
import ConnectionModal from "@scratch-submodule/scratch-gui/src/containers/connection-modal.jsx";

import layout, {
  BLOCKS_DEFAULT_SCALE,
  STAGE_SIZE_MODES,
} from "@scratch-submodule/scratch-gui/src/lib/layout-constants";
import {
  resolveStageSize,
  StageSizeMode,
} from "@scratch-submodule/scratch-gui/src/lib/screen-utils";
import {
  ColorTheme,
  themeMap,
} from "@scratch-submodule/scratch-gui/src/lib/themes";

import styles from "@scratch-submodule/scratch-gui/src/components/gui/gui.css";
import crtStyles from "./gui.css";
import addExtensionIcon from "@scratch-submodule/scratch-gui/src/components/gui/icon--extensions.svg";
import codeIcon from "@scratch-submodule/scratch-gui/src/components/gui/icon--code.svg";
import costumesIcon from "@scratch-submodule/scratch-gui/src/components/gui/icon--costumes.svg";
import soundsIcon from "@scratch-submodule/scratch-gui/src/components/gui/icon--sounds.svg";
import { useEffect } from "react";

import Blocks from "../../../containers/customized-scratch-containers/Blocks";
import TargetPane from "../../../containers/customized-scratch-containers/TargetPane";
import StageWrapper from "../stage-wrapper/StageWrapper";

const messages = defineMessages({
  addExtension: {
    id: "gui.gui.addExtension",
    description: "Button to add an extension in the target pane",
    defaultMessage: "Add Extension",
  },
});

// Cache this value to only retrieve it once the first time.
// Assume that it doesn't change for a session.
let isRendererSupported: boolean | null = null;

const GUIComponent = (props: {
  // required
  vm: VM;
  intl: InjectedIntl;

  // optional
  assetHost?: string;
  cloudHost?: string;
  fetchingProject?: boolean;
  isLoading?: boolean;
  isShowingProject?: boolean;
  loadingStateVisible?: boolean;
  onProjectLoaded?: () => void;
  onStorageInit?: () => void;
  onUpdateProjectId?: () => void;
  onVmInit?: () => void;
  projectHost?: string;

  alertsVisible?: boolean;
  connectionModalVisible?: boolean;
  activeTabIndex?: number;
  backdropLibraryVisible?: boolean;
  backpackHost?: string | null;
  backpackVisible?: boolean;
  basePath: string;
  blocksTabVisible?: boolean;
  blocksId?: string;
  canEditTask?: boolean;
  isCostumesTabEnabled?: boolean;
  isSoundsTabEnabled?: boolean;
  cardsVisible?: boolean;
  costumeLibraryVisible?: boolean;
  costumesTabVisible?: boolean;
  isCreating: boolean;
  isFullScreen: boolean;
  isPlayerOnly?: boolean;
  isRtl: boolean;
  loading: boolean;
  logo?: string;
  onActivateCostumesTab?: () => void;
  onActivateSoundsTab?: () => void;
  onActivateTab?: () => void;
  onExtensionButtonClick?: () => void;
  onRequestCloseBackdropLibrary?: () => void;
  onRequestCloseCostumeLibrary?: () => void;
  onShowPrivacyPolicy?: () => void;
  onTabSelect?: () => void;
  onTaskProgress?: (totalAssertions: number, passedAssertions: number) => void;
  soundsTabVisible?: boolean;
  // see https://github.com/scratchfoundation/scratch-gui/blob/d678d609e182ccc5ab557d7d45a3cc3e6430b056/src/lib/layout-constants.js#L7
  stageSizeMode: StageSizeMode;
  targetIsStage?: boolean;
  theme: ColorTheme;
  tipsLibraryVisible?: boolean;
}) => {
  const {
    activeTabIndex,
    alertsVisible,
    basePath,
    backdropLibraryVisible,
    backpackHost,
    backpackVisible,
    blocksId,
    blocksTabVisible,
    cardsVisible,
    canEditTask,
    isCostumesTabEnabled,
    isSoundsTabEnabled,
    connectionModalVisible,
    costumeLibraryVisible,
    costumesTabVisible,
    intl,
    isCreating,
    isFullScreen,
    isPlayerOnly,
    isRtl,
    loading,
    onActivateCostumesTab,
    onActivateSoundsTab,
    onActivateTab,
    onExtensionButtonClick,
    onRequestCloseBackdropLibrary,
    onRequestCloseCostumeLibrary,
    soundsTabVisible,
    stageSizeMode,
    targetIsStage,
    theme,
    tipsLibraryVisible,
    vm,
    onTaskProgress,
    ...componentProps
  } = omit(props, "dispatch");

  const tabClassNames = {
    tabs: styles.tabs,
    tab: classNames(tabStyles.reactTabsTab, styles.tab),
    tabList: classNames(tabStyles.reactTabsTabList, styles.tabList),
    tabPanel: classNames(tabStyles.reactTabsTabPanel, styles.tabPanel),
    tabPanelSelected: classNames(
      tabStyles.reactTabsTabPanelSelected,
      styles.isSelected,
    ),
    tabSelected: classNames(tabStyles.reactTabsTabSelected, styles.isSelected),
  };

  if (isRendererSupported === null) {
    isRendererSupported = Renderer.isSupported();
  }

  useEffect(() => {
    const enableAssertions = () => vm.runtime.emit("ENABLE_ASSERTIONS");

    if (!canEditTask) {
      // if we cannot edit the task, we always enable assertions
      // if the extension is loaded

      const onAssertionsChecked = (
        totalAssertions: number,
        passedAssertions: number,
      ) => onTaskProgress && onTaskProgress(totalAssertions, passedAssertions);

      // enable assertions as soon as the extension is loaded
      vm.runtime.on("ASSERTIONS_EXTENSION_LOADED", enableAssertions);
      vm.runtime.on("ASSERTIONS_CHECKED", onAssertionsChecked);

      // or if the extension is already loaded, enable them now
      enableAssertions();

      return () => {
        // when effect is cleaned up, remove event listeners
        vm.runtime.removeListener(
          "ASSERTIONS_EXTENSION_LOADED",
          enableAssertions,
        );

        vm.runtime.off("ASSERTIONS_CHECKED", onAssertionsChecked);

        // and disable assertions that were enabled automatically
        vm.runtime.emit("DISABLE_ASSERTIONS");
      };
    }

    // otherwise we only enable assertions if the user has enabled them
    // return no-op cleanup function
    return () => {};
  }, [vm.runtime, canEditTask, onTaskProgress]);

  return (
    <MediaQuery minWidth={layout.fullSizeMinWidth}>
      {(isFullSize) => {
        const stageSize = resolveStageSize(stageSizeMode, isFullSize);

        return isPlayerOnly ? (
          <StageWrapper
            isFullScreen={isFullScreen}
            isRendererSupported={isRendererSupported || false}
            isRtl={isRtl}
            loading={loading}
            stageSize={STAGE_SIZE_MODES.large}
            vm={vm}
            canEditTask={canEditTask}
          />
        ) : (
          <Box
            className={styles.pageWrapper}
            dir={isRtl ? "rtl" : "ltr"}
            {...componentProps}
          >
            {loading ? <Loader /> : null}
            {isCreating ? <Loader messageId="gui.loader.creating" /> : null}
            {isRendererSupported ? null : <WebGlModal isRtl={isRtl} />}
            {tipsLibraryVisible ? <TipsLibrary /> : null}
            {cardsVisible ? <Cards /> : null}
            {alertsVisible ? (
              <Alerts className={styles.alertsContainer} />
            ) : null}
            {connectionModalVisible ? <ConnectionModal vm={vm} /> : null}
            {costumeLibraryVisible ? (
              <CostumeLibrary
                vm={vm}
                onRequestClose={onRequestCloseCostumeLibrary}
              />
            ) : null}
            {backdropLibraryVisible ? (
              <BackdropLibrary
                vm={vm}
                onRequestClose={onRequestCloseBackdropLibrary}
              />
            ) : null}
            <Box className={crtStyles.bodyWrapper}>
              <Box className={styles.flexWrapper}>
                <Box className={styles.editorWrapper}>
                  <Tabs
                    forceRenderTabPanel
                    className={tabClassNames.tabs}
                    selectedIndex={activeTabIndex}
                    selectedTabClassName={tabClassNames.tabSelected}
                    selectedTabPanelClassName={tabClassNames.tabPanelSelected}
                    onSelect={onActivateTab}
                  >
                    <TabList className={tabClassNames.tabList}>
                      <Tab className={tabClassNames.tab}>
                        <img draggable={false} src={codeIcon} />
                        <FormattedMessage
                          defaultMessage="Code"
                          description="Button to get to the code panel"
                          id="gui.gui.codeTab"
                        />
                      </Tab>
                      {isCostumesTabEnabled && (
                        <Tab
                          className={tabClassNames.tab}
                          onClick={onActivateCostumesTab}
                        >
                          <img draggable={false} src={costumesIcon} />
                          {targetIsStage ? (
                            <FormattedMessage
                              defaultMessage="Backdrops"
                              description="Button to get to the backdrops panel"
                              id="gui.gui.backdropsTab"
                            />
                          ) : (
                            <FormattedMessage
                              defaultMessage="Costumes"
                              description="Button to get to the costumes panel"
                              id="gui.gui.costumesTab"
                            />
                          )}
                        </Tab>
                      )}
                      {isSoundsTabEnabled && (
                        <Tab
                          className={tabClassNames.tab}
                          onClick={onActivateSoundsTab}
                        >
                          <img draggable={false} src={soundsIcon} />
                          <FormattedMessage
                            defaultMessage="Sounds"
                            description="Button to get to the sounds panel"
                            id="gui.gui.soundsTab"
                          />
                        </Tab>
                      )}
                    </TabList>
                    <TabPanel className={tabClassNames.tabPanel}>
                      <Box className={styles.blocksWrapper}>
                        <Blocks
                          canEditTask={canEditTask}
                          key={`${blocksId}/${theme}`}
                          grow={1}
                          isVisible={blocksTabVisible}
                          options={{
                            media: `${basePath}static/${themeMap[theme].blocksMediaFolder}/`,
                            zoom: {
                              controls: true,
                              wheel: true,
                              startScale: BLOCKS_DEFAULT_SCALE,
                            },
                            grid: {
                              spacing: 40,
                              length: 2,
                              colour: "#ddd",
                            },
                            comments: true,
                            collapse: false,
                            sounds: false,
                          }}
                          stageSize={stageSize}
                          theme={theme}
                          vm={vm}
                        />
                      </Box>
                      {canEditTask && (
                        <Box className={styles.extensionButtonContainer}>
                          <button
                            className={styles.extensionButton}
                            title={intl.formatMessage(messages.addExtension)}
                            onClick={onExtensionButtonClick}
                            data-testid="add-extension-button"
                          >
                            <img
                              className={styles.extensionButtonIcon}
                              draggable={false}
                              src={addExtensionIcon}
                            />
                          </button>
                        </Box>
                      )}
                      <Box className={styles.watermark}>
                        <Watermark />
                      </Box>
                    </TabPanel>
                    {isCostumesTabEnabled && (
                      <TabPanel className={tabClassNames.tabPanel}>
                        {costumesTabVisible ? <CostumeTab vm={vm} /> : null}
                      </TabPanel>
                    )}
                    {isSoundsTabEnabled && (
                      <TabPanel className={tabClassNames.tabPanel}>
                        {soundsTabVisible ? <SoundTab vm={vm} /> : null}
                      </TabPanel>
                    )}
                  </Tabs>
                  {backpackVisible ? <Backpack host={backpackHost} /> : null}
                </Box>

                <Box
                  className={classNames(
                    styles.stageAndTargetWrapper,
                    styles[stageSize],
                  )}
                >
                  <StageWrapper
                    isFullScreen={isFullScreen}
                    isRendererSupported={isRendererSupported || false}
                    isRtl={isRtl}
                    isStageInteractive={canEditTask}
                    stageSize={stageSize}
                    vm={vm}
                    canEditTask={canEditTask}
                  />
                  <Box className={styles.targetWrapper}>
                    <TargetPane
                      isStageSelectorVisible={canEditTask}
                      isAddNewSpriteButtonVisible={canEditTask}
                      isSpriteInfoEnabled={canEditTask}
                      isDeleteSpriteButtonVisible={canEditTask}
                      isDuplicateSpriteButtonVisible={canEditTask}
                      isExportSpriteButtonVisible={canEditTask}
                      stageSize={stageSize}
                      vm={vm}
                    />
                  </Box>
                </Box>
              </Box>
            </Box>
            <DragLayer />
          </Box>
        );
      }}
    </MediaQuery>
  );
};

const mapStateToProps = (state: {
  scratchGui: {
    timeTravel: {
      year: number;
    };
    stageSize: {
      stageSize: StageSizeMode;
    };
    theme: {
      theme: ColorTheme;
    };
  };
}) => ({
  // This is the button's mode, as opposed to the actual current state
  blocksId: state.scratchGui.timeTravel.year.toString(),
  stageSizeMode: state.scratchGui.stageSize.stageSize,
  theme: state.scratchGui.theme.theme,
});

export default injectIntl(connect(mapStateToProps)(GUIComponent));
