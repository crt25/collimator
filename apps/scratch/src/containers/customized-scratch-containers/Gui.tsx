import React from "react";
import { Action, compose, Dispatch } from "redux";
import { connect } from "react-redux";
import VM from "scratch-vm";
import { InjectedIntl, injectIntl } from "react-intl";

import GUIComponent from "../../components/customized-scratch-components/gui/Gui";
import ErrorBoundaryHOC from "@scratch-submodule/scratch-gui/src/lib/error-boundary-hoc.jsx";
import {
  getIsError,
  getIsShowingProject,
  onFetchedProjectData,
} from "@scratch-submodule/scratch-gui/src/reducers/project-state";
import {
  activateTab,
  BLOCKS_TAB_INDEX,
  COSTUMES_TAB_INDEX,
  SOUNDS_TAB_INDEX,
} from "@scratch-submodule/scratch-gui/src/reducers/editor-tab";

import {
  closeCostumeLibrary,
  closeBackdropLibrary,
  openExtensionLibrary,
} from "@scratch-submodule/scratch-gui/src/reducers/modals";

import FontLoaderHOC from "@scratch-submodule/scratch-gui/src/lib/font-loader-hoc.jsx";
import LocalizationHOC from "@scratch-submodule/scratch-gui/src/lib/localization-hoc.jsx";
import TitledHOC from "@scratch-submodule/scratch-gui/src/lib/titled-hoc.jsx";
import storage from "@scratch-submodule/scratch-gui/src/lib/storage";
import vmListenerHOC from "@scratch-submodule/scratch-gui/src/lib/vm-listener-hoc.jsx";
import vmManagerHOC from "@scratch-submodule/scratch-gui/src/lib/vm-manager-hoc.jsx";
import systemPreferencesHOC from "@scratch-submodule/scratch-gui/src/lib/system-preferences-hoc.jsx";

import { setIsScratchDesktop } from "@scratch-submodule/scratch-gui/src/lib/isScratchDesktop.js";
import { StageSizeMode } from "@scratch-submodule/scratch-gui/src/lib/screen-utils";
import { AppStateHOC } from "@scratch-submodule/scratch-gui/src";
import HashParserHOC from "@scratch-submodule/scratch-gui/src/lib/hash-parser-hoc";
import { loadCrtProject } from "../../vm/load-crt-project";

const { RequestMetadata, setMetadata, unsetMetadata } = storage.scratchFetch;

const setProjectIdMetadata = (projectId: string | number) => {
  // If project ID is '0' or zero, it's not a real project ID. In that case, remove the project ID metadata.
  // Same if it's null undefined.
  if (projectId && projectId !== "0") {
    setMetadata(RequestMetadata.ProjectId, projectId);
  } else {
    unsetMetadata(RequestMetadata.ProjectId);
  }
};

interface Props {
  vm: VM;
  assetHost?: string;
  children?: React.ReactNode;
  cloudHost?: string;
  error?: unknown | string;
  fetchingProject?: boolean;
  intl: InjectedIntl;
  isError?: boolean;
  isLoading?: boolean;
  isScratchDesktop: boolean;
  isShowingProject?: boolean;
  isTotallyNormal: boolean;
  isCreatingNew: boolean;
  loadingStateVisible?: boolean;
  onProjectLoaded?: () => void;
  onSeeCommunity?: () => void;
  onStorageInit?: (storage: unknown) => void;
  onUpdateProjectId?: (projectId: string | number) => void;
  onVmInit?: (vm: VM) => void;
  projectHost?: string;
  projectId: string | number;
  isRtl: boolean;
  isFullScreen: boolean;
  basePath: string;

  onFetchedProjectData: (projectData: unknown, loadingState: unknown) => void;

  backpackHost: string | null;
  backpackVisible: boolean;
  blocksId: string;
  canChangeLanguage: boolean;
  canChangeTheme: boolean;
  canCreateNew: boolean;
  canEditTitle: boolean;
  canCreateCopy: boolean;
  isCreating: boolean;
  stageSizeMode: StageSizeMode;
}

interface ReduxState {
  scratchGui: {
    projectState: {
      loadingState: string;
      error: unknown | string;
      projectId: string | number;
    };
    editorTab: {
      activeTabIndex: number;
    };
    alerts: {
      visible: boolean;
    };
    cards: {
      visible: boolean;
    };
    mode: {
      isFullScreen: boolean;
      isPlayerOnly: boolean;
    };
    modals: {
      backdropLibrary: boolean;
      connectionModal: boolean;
      costumeLibrary: boolean;
      loadingProject: boolean;
      tipsLibrary: boolean;
    };
    targets: {
      stage: {
        id: string;
      };
      editingTarget: string;
    };
    vm: VM;
  };
  locales: {
    isRtl: boolean;
  };
}

class GUI extends React.Component<Props> {
  componentDidMount() {
    setIsScratchDesktop(this.props.isScratchDesktop);

    if (this.props.onStorageInit) {
      this.props.onStorageInit(storage);
    }

    if (this.props.onVmInit) {
      this.props.onVmInit(this.props.vm);
    }

    setProjectIdMetadata(this.props.projectId);

    storage.setAssetHost("https://assets.scratch.mit.edu");
    storage.setTranslatorFunction(this.props.intl.formatMessage);
    storage
      .load(storage.AssetType.Project, "0", storage.DataFormat.JSON)
      .then((projectAsset) => {
        if (projectAsset) {
          loadCrtProject(
            this.props.vm,
            (projectAsset as unknown as { data: object }).data,
          );
        } else {
          throw new Error("Could not load default project");
        }
      })
      .catch((err) => {
        console.error(err);
      });
  }

  componentDidUpdate(prevProps: Props) {
    if (this.props.projectId !== prevProps.projectId) {
      if (this.props.projectId !== null && this.props.onUpdateProjectId) {
        this.props.onUpdateProjectId(this.props.projectId);
      }
      setProjectIdMetadata(this.props.projectId);
    }
    if (
      this.props.isShowingProject &&
      !prevProps.isShowingProject &&
      this.props.onProjectLoaded
    ) {
      // this only notifies container when a project changes from not yet loaded to loaded
      // At this time the project view in www doesn't need to know when a project is unloaded
      this.props.onProjectLoaded();
    }
  }

  render() {
    if (this.props.isError) {
      throw new Error(
        `Error in Scratch GUI [location=${window.location}]: ${this.props.error}`,
      );
    }
    const {
      /* eslint-disable @typescript-eslint/no-unused-vars */
      assetHost,
      cloudHost,
      error,
      isError,
      isScratchDesktop,
      isShowingProject,
      onProjectLoaded,
      onStorageInit,
      onUpdateProjectId,
      onVmInit,
      projectHost,
      projectId,
      isCreatingNew,
      onFetchedProjectData,
      /* eslint-enable @typescript-eslint/no-unused-vars */
      children,
      fetchingProject,
      isLoading,
      loadingStateVisible,
      ...componentProps
    } = this.props;
    return (
      <GUIComponent
        loading={fetchingProject || isLoading || loadingStateVisible || false}
        {...componentProps}
      >
        {children}
      </GUIComponent>
    );
  }
}

const mapStateToProps = (state: ReduxState) => {
  const loadingState = state.scratchGui.projectState.loadingState;
  return {
    activeTabIndex: state.scratchGui.editorTab.activeTabIndex,
    alertsVisible: state.scratchGui.alerts.visible,
    backdropLibraryVisible: state.scratchGui.modals.backdropLibrary,
    blocksTabVisible:
      state.scratchGui.editorTab.activeTabIndex === BLOCKS_TAB_INDEX,
    cardsVisible: state.scratchGui.cards.visible,
    connectionModalVisible: state.scratchGui.modals.connectionModal,
    costumeLibraryVisible: state.scratchGui.modals.costumeLibrary,
    costumesTabVisible:
      state.scratchGui.editorTab.activeTabIndex === COSTUMES_TAB_INDEX,
    error: state.scratchGui.projectState.error,
    isError: getIsError(loadingState),
    isFullScreen: state.scratchGui.mode.isFullScreen,
    isPlayerOnly: state.scratchGui.mode.isPlayerOnly,
    isRtl: state.locales.isRtl,
    isShowingProject: getIsShowingProject(loadingState),
    loadingStateVisible: state.scratchGui.modals.loadingProject,
    projectId: state.scratchGui.projectState.projectId,
    soundsTabVisible:
      state.scratchGui.editorTab.activeTabIndex === SOUNDS_TAB_INDEX,
    targetIsStage:
      state.scratchGui.targets.stage &&
      state.scratchGui.targets.stage.id ===
        state.scratchGui.targets.editingTarget,
    tipsLibraryVisible: state.scratchGui.modals.tipsLibrary,
    vm: state.scratchGui.vm,
  };
};

const mapDispatchToProps = (dispatch: Dispatch<Action>) => ({
  onExtensionButtonClick: () => dispatch(openExtensionLibrary()),
  onActivateTab: (tab: number) => dispatch(activateTab(tab)),
  onActivateCostumesTab: () => dispatch(activateTab(COSTUMES_TAB_INDEX)),
  onActivateSoundsTab: () => dispatch(activateTab(SOUNDS_TAB_INDEX)),
  onRequestCloseBackdropLibrary: () => dispatch(closeBackdropLibrary()),
  onRequestCloseCostumeLibrary: () => dispatch(closeCostumeLibrary()),
  onFetchedProjectData: (projectData: unknown, loadingState: unknown) =>
    dispatch(onFetchedProjectData(projectData, loadingState) as Action),
});

const ConnectedGUI = injectIntl(
  connect(mapStateToProps, mapDispatchToProps)(GUI),
);

// note that redux's 'compose' function is just being used as a general utility to make
// the hierarchy of HOC constructor calls clearer here; it has nothing to do with redux's
// ability to compose reducers.
const WrappedGui = compose(
  LocalizationHOC,
  ErrorBoundaryHOC("Top Level App"),
  FontLoaderHOC,
  TitledHOC,
  vmListenerHOC,
  vmManagerHOC,
  systemPreferencesHOC,
)(ConnectedGUI);

// Analogous to https://github.com/scratchfoundation/scratch-gui/blob/develop/src/playground/render-gui.jsx#L37
const DoubleWrapedGui = AppStateHOC(HashParserHOC(WrappedGui));

export default DoubleWrapedGui;
