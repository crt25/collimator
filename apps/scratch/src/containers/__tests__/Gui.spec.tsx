import { render } from "@testing-library/react";
import Gui from "../Gui";

describe("Scratch GUI Component", () => {
  it("renders the Scratch UI unchanged", () => {
    const { container } = render(<Gui />);
    expect(container).toMatchSnapshot();
  });
});
