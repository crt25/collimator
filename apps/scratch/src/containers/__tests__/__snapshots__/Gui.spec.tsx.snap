// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Scratch GUI Component renders the Scratch UI unchanged 1`] = `
<div>
  <div
    class=""
    dir="ltr"
    username=""
  >
    <div
      class=""
    >
      <div>
        <div>
          <img
            src="top-block.svg"
          />
          <img
            src="middle-block.svg"
          />
          <img
            src="bottom-block.svg"
          />
        </div>
        <div>
          <span>
            Loading Project
          </span>
        </div>
        <div>
          <div
            style="transform: translate(0, -50px);"
          >
            <div>
              <span>
                Creating blocks …
              </span>
            </div>
            <div>
              <span>
                Loading sprites …
              </span>
            </div>
            <div>
              <span>
                Loading sounds …
              </span>
            </div>
            <div>
              <span>
                Loading extensions …
              </span>
            </div>
            <div>
              <span>
                Creating blocks …
              </span>
            </div>
            <div>
              <span>
                Herding cats …
              </span>
            </div>
            <div>
              <span>
                Transmitting nanos …
              </span>
            </div>
            <div>
              <span>
                Inflating gobos …
              </span>
            </div>
            <div>
              <span>
                Preparing emojis …
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class=""
    >
      <div>
        <div>
          <img
            src="top-block.svg"
          />
          <img
            src="middle-block.svg"
          />
          <img
            src="bottom-block.svg"
          />
        </div>
        <div>
          <span>
            Creating Project
          </span>
        </div>
        <div>
          <div
            style="transform: translate(0, -75px);"
          >
            <div>
              <span>
                Creating blocks …
              </span>
            </div>
            <div>
              <span>
                Loading sprites …
              </span>
            </div>
            <div>
              <span>
                Loading sounds …
              </span>
            </div>
            <div>
              <span>
                Loading extensions …
              </span>
            </div>
            <div>
              <span>
                Creating blocks …
              </span>
            </div>
            <div>
              <span>
                Herding cats …
              </span>
            </div>
            <div>
              <span>
                Transmitting nanos …
              </span>
            </div>
            <div>
              <span>
                Inflating gobos …
              </span>
            </div>
            <div>
              <span>
                Preparing emojis …
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      bounds="parent"
      class=""
    >
      <div
        class=""
      />
    </div>
    <div
      class=""
    >
      <div
        class=""
      >
        <div
          class=""
        >
          <div
            class="react-tabs"
            data-tabs="true"
          >
            <ul
              class=""
              role="tablist"
            >
              <li
                aria-controls="react-tabs-1"
                aria-disabled="false"
                aria-selected="true"
                class="react-tabs__tab--selected"
                id="react-tabs-0"
                role="tab"
                tabindex="0"
              >
                <img
                  draggable="false"
                  src="icon--code.svg"
                />
                <span>
                  Code
                </span>
              </li>
              <li
                aria-controls="react-tabs-3"
                aria-disabled="false"
                aria-selected="false"
                class=""
                id="react-tabs-2"
                role="tab"
              >
                <img
                  draggable="false"
                  src="icon--costumes.svg"
                />
                <span>
                  Backdrops
                </span>
              </li>
              <li
                aria-controls="react-tabs-5"
                aria-disabled="false"
                aria-selected="false"
                class=""
                id="react-tabs-4"
                role="tab"
              >
                <img
                  draggable="false"
                  src="icon--sounds.svg"
                />
                <span>
                  Sounds
                </span>
              </li>
            </ul>
            <div
              aria-labelledby="react-tabs-0"
              class="react-tabs__tab-panel--selected"
              id="react-tabs-1"
              role="tabpanel"
            >
              <div
                class=""
              >
                <div>
                  Blocks
                </div>
              </div>
              <div
                class=""
              >
                <button
                  title="Add Extension"
                >
                  <img
                    draggable="false"
                    src="icon--extensions.svg"
                  />
                </button>
              </div>
              <div
                class=""
              >
                <img />
              </div>
            </div>
            <div
              aria-labelledby="react-tabs-2"
              class=""
              id="react-tabs-3"
              role="tabpanel"
            />
            <div
              aria-labelledby="react-tabs-4"
              class=""
              id="react-tabs-5"
              role="tabpanel"
            />
          </div>
        </div>
        <div
          class=""
        >
          <div
            class=""
            dir="ltr"
          >
            <div
              class=""
            >
              <div
                class=""
              >
                <div
                  class=""
                >
                  <img
                    class=""
                    draggable="false"
                    src="icon--green-flag.svg"
                    title="Go"
                  />
                  <img
                    class=""
                    draggable="false"
                    src="icon--stop-all.svg"
                    title="Stop"
                  />
                </div>
                <div>
                  <div>
                    <div
                      class=""
                    >
                      <button
                        aria-label="Switch to small stage"
                        aria-pressed="false"
                        title="Switch to small stage"
                      >
                        <img
                          aria-hidden="true"
                          src="icon--small-stage.svg"
                        />
                      </button>
                      <button
                        aria-label="Switch to large stage"
                        aria-pressed="true"
                        title="Switch to large stage"
                      >
                        <img
                          aria-hidden="true"
                          src="icon--large-stage.svg"
                        />
                      </button>
                    </div>
                  </div>
                  <div>
                    <span
                      class=""
                      role="button"
                    >
                      <div>
                        <img
                          alt="Enter full screen mode"
                          draggable="false"
                          src="icon--fullscreen.svg"
                          title="Full Screen Control"
                        />
                      </div>
                    </span>
                  </div>
                </div>
              </div>
            </div>
            <div
              class=""
            />
          </div>
          <div
            class=""
          >
            <div
              intl="[object Object]"
            >
              <div
                class=""
              >
                <div
                  class="css-1vud2rv"
                >
                  <div
                    class=""
                  >
                    <div
                      class=""
                    >
                      <div>
                        <label>
                          <span>
                            <span>
                              Sprite
                            </span>
                          </span>
                          <input
                            class=""
                            disabled=""
                            placeholder="Name"
                            tabindex="0"
                            type="text"
                            value=""
                          />
                        </label>
                      </div>
                      <div>
                        <label>
                          <span>
                            x
                          </span>
                          <input
                            class="undefined"
                            disabled=""
                            placeholder="x"
                            tabindex="0"
                            type="text"
                            value=""
                          />
                        </label>
                      </div>
                      <div>
                        <label>
                          <span>
                            y
                          </span>
                          <input
                            class="undefined"
                            disabled=""
                            placeholder="y"
                            tabindex="0"
                            type="text"
                            value=""
                          />
                        </label>
                      </div>
                    </div>
                    <div
                      class=""
                    >
                      <div>
                        <div
                          class="undefined"
                        >
                          <button
                            aria-label="Show sprite"
                            disabled=""
                            title="Show sprite"
                          >
                            <img
                              aria-hidden="true"
                              src="icon--show.svg"
                            />
                          </button>
                          <button
                            aria-label="Hide sprite"
                            aria-pressed="false"
                            disabled=""
                            title="Hide sprite"
                          >
                            <img
                              aria-hidden="true"
                              src="icon--hide.svg"
                            />
                          </button>
                        </div>
                      </div>
                      <div
                        class=""
                      >
                        <label>
                          <span>
                            <span>
                              Size
                            </span>
                          </span>
                          <input
                            class="undefined"
                            disabled=""
                            label="[object Object]"
                            tabindex="0"
                            type="text"
                            value=""
                          />
                        </label>
                      </div>
                      <div
                        class=""
                      >
                        <label>
                          <span>
                            <span>
                              Direction
                            </span>
                          </span>
                          <input
                            class="undefined"
                            disabled=""
                            label="[object Object]"
                            tabindex="0"
                            type="text"
                            value=""
                          />
                        </label>
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  class=""
                >
                  <div
                    class=""
                  />
                </div>
              </div>
              <div>
                <div />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
